src/package.ts

```diff
@@ -30,28 +33,181 @@ export interface Package {
+  );
+  const tagNameMatchingLatestRelease = expectedReleaseTagNames.find(
+    (tagName) => {
+      return projectTagNames.includes(tagName);
+    },
+  );
+
+  if (
+    projectTagNames.length > 0 &&
+    tagNameMatchingLatestRelease === undefined
+  ) {
+    throw new Error(
+      format(
+        'The package %s has no Git tag for its current version %s (expected %s), so this tool is unable to determine whether it should be included in this release. You will need to create a tag for this package in order to proceed.',
+        validatedManifest.name,
+        validatedManifest.version,
+        expectedReleaseTagNames.map((tagName) => `"${tagName}"`).join(', '),
+      ),
+    );
+  }
+
+  const hasChangesSinceLatestRelease =
+    tagNameMatchingLatestRelease === undefined
+      ? true
+      : await hasChangesInDirectorySinceGitTag(
+          projectDirectoryPath,
+          packageDirectoryPath,
+          tagNameMatchingLatestRelease,
+        );
+
+  return {
+    directoryPath: packageDirectoryPath,
+    manifestPath,
+    validatedManifest,
+    unvalidatedManifest,
+    changelogPath,
+    hasChangesSinceLatestRelease,
+  };
+}
+
+/**
+ * Collects information about a workspace package within a monorepo.
+ *
+ * @param args - The arguments to this function.
+ * @param args.packageDirectoryPath - The path to a package within a project.
+ * @param args.rootPackageName - The name of the root package within the
+ * monorepo to which this package belongs.
+ * @param args.rootPackageVersion - The version of the root package within the
+ * monorepo to which this package belongs.
+ * @param args.projectDirectoryPath - The path to the project directory.
+ * @param args.projectTagNames - The tag names across the whole project.
  * @returns Information about the package.
  */
-export async function readPackage(
-  packageDirectoryPath: string,
-): Promise<Package> {
+export async function readMonorepoWorkspacePackage({
+  packageDirectoryPath,
+  rootPackageName,
+  rootPackageVersion,
+  projectDirectoryPath,
+  projectTagNames,
+}: {
+  packageDirectoryPath: string;
+  rootPackageName: string;
+  rootPackageVersion: SemVer;
+  projectDirectoryPath: string;
+  projectTagNames: string[];
+}): Promise<Package> {
   const manifestPath = path.join(packageDirectoryPath, MANIFEST_FILE_NAME);
   const changelogPath = path.join(packageDirectoryPath, CHANGELOG_FILE_NAME);
   const { unvalidated: unvalidatedManifest, validated: validatedManifest } =
     await readPackageManifest(manifestPath);
+  const expectedWorkspacePackageReleaseTagNames =
+    generateMonorepoWorkspacePackageReleaseTagNames(
+      validatedManifest.name,
+      validatedManifest.version.toString(),
+    );
+  const expectedRootPackageReleaseTagNames =
+    generateMonorepoRootPackageReleaseTagNames(rootPackageVersion.toString());
+  const tagNameMatchingLatestRelease = [
+    ...expectedWorkspacePackageReleaseTagNames,
+    ...expectedRootPackageReleaseTagNames,
```

Comment:
Perhaps it's worth printing a console warning if the fallback root tag is used? This should only happen in rare circumstances, like when using this tool for the first time on a monorepo that didn't previously have workspace tags (or at least not in the expected format).

A warning might hint that something is amiss if this happens outside of that scenario.

Reply:
If we display a message to the user, I'd like it to be helpful and include a suggestion as to how they can resolve the problem, but I'm curious what that suggestion would be, because I think it differs based on whether this is the first time this tool is being used on a monorepo or not. If it is not the first time, then we can presume that all packages have been tagged properly, and so if the current release for any given package does _not_ have the tag we expect, that is an anomaly and should be corrected. However, if it _is_ the first time, then we can presume that no releases for any of the packages in the monorepo have tags at all, so in that case, if we say that the current release for a given package needs to be tagged properly, we might as well say that ALL releases for a given package needs to be tagged properly.

So, either we need a way to tell whether this is the first time a user is using this tool, or we need to offer another tool the user can use which will validate that all releases for all packages have proper tags and backfill them if necessary.

Thoughts/

Reply:
Originally I was thinking the user would just want to know the fallback occurred, since as you point out it's not necessarily a problem or something that is actionable.

If we can do better, that would be great. I'll give this some thought.

Reply:
>If it is not the first time, then we can presume that all packages have been tagged properly, and so if the current release for any given package does not have the tag we expect, that is an anomaly and should be corrected.

Yes, agreed. We could even throw an error in this case, this should never happen. That would be an improvement. It's distinct from the suggestion I meant to make here though.

>However, if it is the first time, then we can presume that no releases for any of the packages in the monorepo have tags at all, so in that case, if we say that the current release for a given package needs to be tagged properly, we might as well say that ALL releases for a given package need to be tagged properly.

If we're just starting to use this tool, I don't think we can make any assumptions about tags. We don't know what convention, if any, they were using before.

To be clear, I was suggesting that we warn in the case where we find a root monorepo tag but not a workspace tag. We proceed silently in that circumstance in this PR at the moment.

This seems worth highlighting so that the user understands it's happening, even just for auditability. For example, if there were unpublished changes in a workspace that predate the most recent root tag, they would be missed because of us using this fallback. I don't think this scenario is likely, nor do I think it makes sense necessarily to force the user to manually add tags. But calling out that it's happening might leave an audit trail to follow up on if it happens and someone is confused as to why.

Reply:
Fair enough! Added warning in [565c2a5](https://github.com/MetaMask/create-release-branch/pull/23/commits/565c2a505879520c811bf593e75a42913590f313).